satpy/modifiers/angles.py

```diff
@@ -395,6 +395,8 @@ def get_cos_sza(data_arr: xr.DataArray) -> xr.DataArray:
     """
     chunks = _geo_chunks_from_data_arr(data_arr)
     lons, lats = _get_valid_lonlats(data_arr.attrs["area"], chunks)
+    lons = lons.astype(data_arr.dtype)
+    lats = lats.astype(data_arr.dtype)
```

Comment:
Should this simply be hardcoded `float32` (or even `float16`)? If those are only used for SZA corrections, I don't see why we would ever need a higher precision, or why the precision should depend on the dtype of the input data.  I don't know if `data_arr.dtype` could ever be integer here, but if it is, that would impact the quality of the correction.

Reply:
I could have sworn I fixed this inside the `_get_cos_sza` call or something. Let me track it down. I don't think I agree with casting lon/lats to the data array dtype here.

Reply:
https://github.com/pytroll/satpy/commit/e6e5c6e5f0085c8c8216951b80a46a468b0f75a2

Reply:
The `sunzen_corr_cos()` isn't used because, well, we're not correcting for cosine angle, just using it to weight the overlap. The casting is needed to keep the data at proper dtype. I wanted to pass the dtype to `area.get_lonlats()`, but the decorated zarr-caching is out of my understanding and just passing the dtype as another argument didn't work.

Reply:
Ah ok.

I don't think the cached functions should be correcting for dtype in this sense. By doing that you'd end up with possibly two cached versions of the data with the only different being 32-bit and 64-bit floating dtype. What about casting the `cos_sza` after the `_get_cos_sza` call? Or is that not good enough?

The `_get_cos_sza` function is also used in the below `_get_sun_angles` function too so we may want this type of casting done there too. Also, isn't this casting dangerous if the input array is an integer type?

Side note: The `get_valid_lonlats` should maybe make sure that `np.nan` is the same dtype as the lon/lats when masking to avoid upcasting there. I think we usually use 64-bit floats for all x/y coordinates so the lon/lats are usually 64-bit too so probably not a huge problem.

Reply:
Casting as it is in this PR, processing time 49.1 s:
![Screenshot 2023-11-20 at 11-04-43 Bokeh Plot](https://github.com/pytroll/satpy/assets/3170788/1149b3f6-3e4a-4067-9914-bbd98a3bbf81)

Casting the result of call to `cos_sza()`, processing time 51.6 s:
![Screenshot 2023-11-20 at 11-06-11 Bokeh Plot](https://github.com/pytroll/satpy/assets/3170788/e1cbb13d-e07f-447b-a0b7-748183388e32)

The latter uses ~1 GB more memory.

Reply:
Oh, and no visible difference for 8-bit imagery.

Reply:
The lons/lats casting is made more strict in https://github.com/pytroll/satpy/pull/2640/commits/e98eb3ee54831dd2449f7fc73a72c5bdff6b4324 . The casting is made only if the dtypes of lons/lats and the data don't match, and data are floats.

Reply:
Ah this seems much safer. Thanks.